imagehash.py

```diff
@@ -372,3 +376,219 @@ def colorhash(image, binbits=3):
+		# In the case of a draw, pick the image with the lower overall hamming distance
+		return other_hashes[min(
+			winner_idxs,
+			key=lambda other_hash_idx: sum(
+				other_segment_hash - segment_hash
+				for segment_hash in self.segment_hashes
+				for other_segment_hash in other_hashes[other_hash_idx].segment_hashes
+			)
+		)]
+
+
+def _find_region(remaining_pixels, segmented_pixels):
+	"""
+	Finds a region and returns a set of pixel coordinates for it.
+	:param remaining_pixels: A numpy bool array, with True meaning the pixels are remaining to segment
+	:param segmented_pixels: A set of pixel coordinates which have already been assigned to segment. This will be
+	updated with the new pixels added to the returned segment.
+	"""
+	in_region = set()
+	not_in_region = set()
+	# Find the first pixel in remaining_pixels with a value of True
+	available_pixels = numpy.transpose(numpy.nonzero(remaining_pixels))
+	start = tuple(available_pixels[0])
+	in_region.add(start)
+	new_pixels = in_region.copy()
+	while True:
+		try_next = set()
+		# Find surrounding pixels
+		for pixel in new_pixels:
+			x, y = pixel
+			neighbours = [
+				(x-1, y),
+				(x+1, y),
+				(x, y-1),
+				(x, y+1)
+			]
+			try_next.update(neighbours)
+		# Remove pixels we have already seen
+		try_next.difference_update(segmented_pixels, not_in_region)
+		# If there's no more pixels to try, the region is complete
+		if not try_next:
+			break
+		# Empty new pixels set, so we know whose neighbour's to check next time
+		new_pixels = set()
+		# Check new pixels
+		for pixel in try_next:
+			if remaining_pixels[pixel]:
+				in_region.add(pixel)
+				new_pixels.add(pixel)
+				segmented_pixels.add(pixel)
+			else:
+				not_in_region.add(pixel)
+	return in_region
+
+
+def segmented_hash(
+		image,
+		hash_func=None,
+		limit_segments=None,
+		segment_threshold=128,
+		min_segment_size=100,
+		segmentation_image_size=300
+	):
+	"""
+	Creates a CropResistantHash object, by the algorithm described in the paper "Efficient Cropping-Resistant Robust
+	Image Hashing". DOI 10.1109/ARES.2014.85
+	This algorithm partitions the image into bright and dark segments, using a watershed-like algorithm, and then does
+	an image hash on each segment. This makes the image much more resistant to cropping than other algorithms, with
+	the paper claiming resistance to up to 50% cropping, while most other algorithms stop at about 5% cropping.
+	:param image: The image to hash
+	:param hash_func: The hashing function to use
+	:param limit_segments: If you have storage requirements, you can limit to hashing only the M largest segments
+	:param segment_threshold: Brightness threshold between hills and valleys. This should be static, putting it between
+	peak and trough dynamically breaks the matching
+	:param min_segment_size: Minimum number of pixels for a hashable segment
+	:param segmentation_image_size: Size which the image is resized to before segmentation
+	"""
+	# Define some tunable params
```

Comment:
They should probably go in the argument list.

Reply:
I wasn't too sure on these actually, as they're the default values for the filters anyway, I'm not sure what a user would gain by tuning them, I added the tunable params comment when I threw the first draft together.

I'm tempted to just remove them, and allow the defaults for the filters to do their thing. Which would you prefer?

Reply:
I'm fine with using defaults and hiding complexity. We can expose those later with the default values if needed.

Reply:
Sounds good, done that